framework/src/kernels/ConservativeAdvection.C

```diff
@@ -59,15 +64,23 @@ ConservativeAdvectionTempl<true>::validParams()
 template <bool is_ad>
 ConservativeAdvectionTempl<is_ad>::ConservativeAdvectionTempl(const InputParameters & parameters)
   : GenericKernel<is_ad>(parameters),
-    _velocity(this->isParamValid("velocity_variable")
-                  ? &this->template coupledGenericVectorValue<is_ad>("velocity_variable")
-                  : (this->isParamValid("velocity_material")
-                         ? &this->template getGenericMaterialProperty<RealVectorValue, is_ad>(
-                                    "velocity_material")
-                                .get()
-                         : nullptr)),
+    _scalar(this->template getGenericMaterialProperty<Real, is_ad>("scalar")),
+    _field_variable(isParamValid("field_variable") ? &this->coupledGradient("field_variable")
```

Comment:
If you used `coupledGenericGradient`, then I believe you should get a `MooseArray<GenericRealVectorValue<is_ad>>` which is the exact same type as your `_velocity`. So I think we should leverage that and not introduce this additional data member (currently named `_field_variable`)